### PR TITLE
Separate dev and pub logs. Prepare devlog for DCE.

### PR DESCRIPTION
**Pull request by dvoytenko**
All, food for thought. One way we can split both logs.

/cc @cramforce @erwinmombay @sriramkrish85 @jridgewell @mkhatib

### CODE REVIEW

File: src/log.js [hunk 1]
```diff
@@ -124,7 +125,109 @@ export class Log {
...
+ * @const {!Log}
+ * DO NOT SUBMIT: always enabled? or always in mode.development?
+ */
+export const publog = new Log(window);
+
+/**
+ * AMP development log. Stripped in the PROD binary.
+ * @const {!Log}
+ */
+export const devlog = new Log(widow);
```
Comment (erwinmombay) [inline src/log.js]: window

**Comment by cramforce** [comment]
This is great in general. Would like to bikeshed the names :)

**Comment by dvoytenko** [comment]
@cramforce Names definitely subject to discuss and change.

How do you feel about moving assert statements into these interfaces? I liked that because then we have a single entry point into pub vs dev separation. I didn't like it because normally `assert != log` and `log.error != throw error`. Also, we currently have a bunch of "assert" statements throughout the codebase, such as `assertHttpsUrl`.  It might be a good compromise still.

**Comment by cramforce** [comment]
Yeah, not a super big fan of that move but it is OK.

I think I'd like to make the convention:

`user.log` `user.assert` and `dev.info` respectively

**Comment by dvoytenko** [comment]
@cramforce 

> user.log user.assert and dev.info respectively

Do you mean, you'd just like to call `publog` as `user` and `devlog` as `dev`? Or would you like to avoid `dev.assert` entirely? We currently do use `assert` a lot as `dev.assert`, i..e as a non-user error. Those cases are fairly safe to DCE as well, but good for development.

**Comment by cramforce** [comment]
I mean't to just rename, yeah.

**Comment by dvoytenko** [comment]
@cramforce Sounds good. Renamed. PTAL to see if it looks better.

**Comment by erwinmombay** [comment]
so on actual erasure for the dist, it is going to just be devlog?

**Comment by dvoytenko** [comment]
Yes, `devlog` (or rather now `dev`) will be erased for the dist.

**Comment by dvoytenko** [comment]
@erwinmombay that may or may not include `dev.assert` as well. But not `dev.throwAsync`. So, it's a partial erasure - is that ok? Maybe we only erase `dev.fine`, `dev.info`, etc?

**Comment by erwinmombay** [comment]
@dvoytenko why would `dev.throwAsync` be treated differently than `dev.assert`? agree with the `dev.fine` and `def.info` statements.

**Comment by dvoytenko** [comment]
@erwinmombay It's not necessary, but typically dev `asserts` are throwaway code - e.g. I'm already expecting that value to be X and I just want to make sure it's the case before I move forward. From that point of view, that can be DCE'd. `dev.throwAsync` is different - I explicitly say that it must be a library error - I just don't want it to block. So, in a way, these APIs are only loosely bound together in these interfaces. The idea is to make the distinction once and optimize things for each specific method.

**Comment by dvoytenko** [comment]
@cramforce @erwinmombay this is ready for review.

File: extensions/amp-access/0.1/amp-access.js [hunk 2]
```diff
@@ -359,17 +359,17 @@ export class AccessService {
...
     }
 
-    log.fine(TAG, 'Start authorization via ', this.config_.authorization);
+    dev.fine(TAG, 'Start authorization via ', this.config_.authorization);
     this.toggleTopClass_('amp-access-loading', true);
     const urlPromise = this.buildUrl_(
         this.config_.authorization, /* useAuthData */ false);
     const promise = urlPromise.then(url => {
-      log.fine(TAG, 'Authorization URL: ', url);
+      devlog.fine(TAG, 'Authorization URL: ', url);
```
Comment (cramforce) [inline extensions/amp-access/0.1/amp-access.js]: dev

File: extensions/amp-access/0.1/amp-access.js [hunk 3]
```diff
@@ -596,7 +595,7 @@ export class AccessService {
     if (this.reportViewPromise_) {
       return this.reportViewPromise_;
     }
-    log.fine(TAG, 'start view monitoring');
+    devlog.fine(TAG, 'start view monitoring');
```
Comment (cramforce) [inline extensions/amp-access/0.1/amp-access.js]: dev. Also elsewhere

File: src/log.js [hunk 4]
```diff
@@ -85,46 +135,245 @@ export class Log {
...
    * @param {string} tag
    * @param {...*} var_args
    * @param {?} opt_error
    */
   error(tag, var_args) {
-    if (this.isEnabled_) {
+    if (this.level_ >= LogLevel.ERROR) {
       this.msg_(tag, 'ERROR', Array.prototype.slice.call(arguments, 1));
+    } else {
+      let error = null;
```
Comment (cramforce) [inline src/log.js]: Love it.

File: src/log.js [hunk 5]
```diff
@@ -85,46 +135,245 @@ export class Log {
...
+ *  1. Not disabled using `#log=0`.
+ *  2. Development mode is enabled via `#development=1` or logging is explicitly
+ *     enabled via `#log=D` where D >= 1.
+ *
+ * @const {!Log}
+ */
+export const user = new Log(window, mode => {
+  const logNum = parseInt(mode.log, 10);
+  if (mode.development || logNum >= 1) {
+    return LogLevel.FINE;
```
Comment (cramforce) [inline src/log.js]: Not super sure this is correct. But it is fine if we in practice only do `user.error` and `user.warn`

Comment (cramforce) [inline src/log.js]: Do you ever see us doing `user.fine`?

File: extensions/amp-access/0.1/amp-access.js [hunk 6]
```diff
@@ -202,20 +202,20 @@ export class AccessService {
 
     // Check that all URLs are valid.
     if (config.authorization) {
-      assertHttpsUrl(config.authorization);
+      user.assertHttpsUrl(config.authorization);
```
Comment (cramforce) [inline extensions/amp-access/0.1/amp-access.js]: Did you actually move this in this CL?

File: src/log.js [hunk 7]
```diff
@@ -85,46 +135,245 @@ export class Log {
...
    */
   warn(tag, var_args) {
-    if (this.isEnabled_) {
+    if (this.level_ >= LogLevel.WARN) {
       this.msg_(tag, 'WARN', Array.prototype.slice.call(arguments, 1));
     }
   }
 
   /**
+   * Reports an error message. If the loggin is disabled, the error is rethrown
```
Comment (erwinmombay) [inline src/log.js]: loggin -> logging?

File: test/functional/test-log.js [hunk 8]
```diff
@@ -0,0 +1,428 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+      Log,
+      LogLevel,
+      USER_ERROR_SENTINEL,
+      dev,
+      isUserErrorMessage,
```
Comment (erwinmombay) [inline test/functional/test-log.js]: too much indent here?

File: extensions/amp-access/0.1/amp-access.js (hunk 2 above)
Comment (dvoytenko) [inline extensions/amp-access/0.1/amp-access.js]: Correct. But I will simply revert this file - I'd rather do replacements globally through the whole codebase. It's here for now as a demo.

File: src/log.js (hunk 7 above)
Comment (dvoytenko) [inline src/log.js]: Done

**Comment by dvoytenko** [comment]
@cramforce @erwinmombay All requests have been fixed. PTAL.

**Comment by cramforce** [comment]
LGTM

**Pull request merged**